Sprint_16

MiraiChan opened the pull request:
Sprint_16 done.

MiraiChan committed:
Onboarding file with the use of UIPageViewController and UIPageControl was added and set up.

MiraiChan committed:
Implemented app's conditional  initialisation based on first launch status.

MiraiChan committed:
Interaction with Core Data was placed in the model layer TrackerCategoryStore.
Computed property 'trackerCategories' for getting data from Core Data and method for storing data( func addNewCategory() ) in Core Data were realised.

MiraiChan committed:
CategoryViewController was set, constraints activated.
ViewModel and UITableViewCell were added.

MiraiChan committed:
ViewModel logic managing TrackerCategory instances was set:
- Methods to add categories and trackers. were Implemented.
- Introduced selectedCategory property with observable pattern.
- Implemented TrackerCategoryStoreDelegate to update categories on changes.
Observable Property Wrapper was implemented:
- Introduced Observable property wrapper to enable observation of property changes.
- Defined wrappedValue, projectedValue, and bind methods.
TrackerCategoryStore was configured accordingly and used to handle category storage and retrieval.
UITableViewCell was configured.

MiraiChan committed:
Updated the code (added instances and properties, deleted excessive unused mock properties) to include a category when appending a new tracker.

MiraiChan committed:
Updated the content of the cells in UITableView.
Made changes to the didSelectRowAt method in a table view delegate:
- added logic to handle the case when indexPath.row is 0, which involves presenting addCategoryViewController and observing changes to the selected category.
Made changes to the appearance of separators in table view cells:
- added logic to conditionally add a separator view only if the current cell is not the last cell in the section.

MiraiChan committed:
Fixed type issue in TrackerCategoryCoreData attribute. Updated code in IrregularEvent file to include category when appending a new tracker.

MiraiChan committed:
Fixed id issue with attribute of entity TrackerCoreData.
Deleted unnecessary lines of code.
Added method to check the state of a createButton in when creating IrregularEvent.

So1ovyov commented on Tracker/Category/CategoryCell.swift:
```diff
@@ -0,0 +1,69 @@
+        setupConstraints()
+    }
+    
+    private func setupSubviews() {
```
В целом можем избавиться от этого метода, а также от повторяющегося `translatesAutoresizingMaskIntoConstraints`
Давай в метод `setupConstraints()` добавим следующее решение с использованием `forEach`
Будет выглядеть вот так:
```
        [titleLabel, doneImage].forEach {
            addSubview($0)
            $0.translatesAutoresizingMaskIntoConstraints = false
        }
```

Таким образом setupSubviews() удаляется, translatesAutoresizingMaskIntoConstraints = false удаляется, да и doneImage можно объявить короче:
`private let doneImage = UIImageView()`

MiraiChan replied:
Done.

So1ovyov commented on Tracker/Category/CategoryViewController.swift:
```diff
@@ -0,0 +1,208 @@
+
+final class CategoryViewController: UIViewController {
+    
+    let cellReuseIdentifier = "HabitCategoryViewController"
```
Кажется не хватает модификатора `private`

MiraiChan replied:
Done.

So1ovyov commented on Tracker/Category/CategoryViewController.swift:
```diff
@@ -0,0 +1,208 @@
+    }
+    
+    private func addSubviews() {
+        view.addSubview(categotyHeader)
```
Также здесь можно использовать `forEach` для `addSubview` и `.translatesAutoresizingMaskIntoConstraints = false`

MiraiChan replied:
Ok.

So1ovyov commented on Tracker/Category/CategoryViewModel.swift:
```diff
@@ -0,0 +1,39 @@
+    }
+    
+    func addCategory(_ toAdd: String) {
+        try! self.categoryStore.addNewCategory(TrackerCategory(title: toAdd, trackers: []))
```
Давай избавимся от force try (try!) любая операция через force может уронить проект в самый неожиданный момент, давай поменяем на try?

MiraiChan replied:
Я неуверенна правильно ли я избавилась от (try!). Нужен будет основательный фидбек.

So1ovyov replied:
Все верно, использовать опциональный try это нормально

So1ovyov commented on Tracker/Category/Observable.swift:
```diff
@@ -0,0 +1,29 @@
+//
+
+@propertyWrapper
+final class Observable<Value> {
```
Отличное решение, но если будет интересно можно пойти в сторону реактивщины и изучить `RxSwift`

So1ovyov commented on Tracker/CoreData/DataStore/TrackerCategoryStore.swift:
```diff
@@ -7,3 +7,108 @@
+    
+    convenience override init() {
+        let context = (UIApplication.shared.delegate as! AppDelegate).context
+        try! self.init(context: context)
```
Давай избавимся от try!

MiraiChan replied:
Легко сказать, то что мне наставники посоветовали это "Как вариант например инициализатор сделать тоже throws и тогда поидее можно будет и try! убрать и вместо return написать throw <SomeError>", но у меня туча ошибок вылазиет, в основном : convenience override init?() { }Failable initializer 'init()' cannot override a non-failable initializer, Cannot override non-throwing initializer with throwing initializer и 'self.init' isn't called on all paths before returning from initializer. Мне кажется, что возможно, проблема в файле AppDelegate.swift может его как-то рефакторить нужно? Пока как временный вариант я что смогла то и сделала с чатом GPT. Если возможно, объясните пути решения проблемы.

So1ovyov replied:
Можно попробовать следующий вариант:
```
    convenience override init() {
        let context = CoreDataBase.shared.context
        self.init(context: context)
    }
```
где CoreDataBase это твой класс в котором ты настраиваешь NSPersistentContainer
```
final class CoreDataBase {

    static let shared = CoreDataBase()
    
    private lazy var persistentContainer: NSPersistentContainer = {
        let container = NSPersistentContainer(name: "TrackerBase")
        container.loadPersistentStores(completionHandler: { (storeDescription, error) in
            if let error = error as NSError? {
                assertionFailure("Не удалось загрузить базу данных!")
            }
        })
        return container
    }()

    var context: NSManagedObjectContext {
        return persistentContainer.viewContext
    }
    
    private init() {}
}
```

MiraiChan replied:
сначала просит поставить обязательно try, а потом 'self.init' isn't called on all paths before returning from initializer The problem is that if the initialization succeeds, there is no corresponding code to execute after the try self.init(context: context) line. You need to add the necessary code to handle the successful initialization or to perform any additional setup. Если писать методы  для Additional setup after successful initialization , то получается громоздко, если так норм , то я могу туда кода понаписать, но надо ли?

So1ovyov commented on Tracker/Models/Tracker.swift:
```diff
@@ -13,4 +13,12 @@ struct Tracker {
     let emoji: String
     let schedule: [TrackerSchedule.DaysOfTheWeek]?
+    
+    init(id: UUID, name: String, color: UIColor, emoji: String, schedule: [TrackerSchedule.DaysOfTheWeek]? = nil) {
```
Было бы круто располагать длинные конструкции так чтобы они легко читались сверху вниз и не растягивались по ширине экрана:
```
    init(
        id: UUID,
        name: String,
        color: UIColor,
        emoji: String,
        schedule: [TrackerSchedule.DaysOfTheWeek]? = nil
    ) {
        self.id = id
        self.name = name
        self.color = color
        self.emoji = emoji
        self.schedule = schedule
    }
```

MiraiChan replied:
Done.

So1ovyov commented on Tracker/TrackersViewController/AddingTracker/NewTrackerViewController/IrregularEvent/IrregularEventViewController.swift:
```diff
@@ -28,6 +34,24 @@ final class IrregularEventViewController: UIViewController {
+    }
+    
+    private var isCategorySelected: Bool {
+        //TODO: add the logic to check if a category is selected
```
Хорошая привычка, молодец!

MiraiChan replied:
Спасибо!

So1ovyov commented on Tracker/TrackersViewController/AddingTracker/NewTrackerViewController/IrregularEvent/IrregularEventViewController.swift:
```diff
@@ -193,6 +219,16 @@ final class IrregularEventViewController: UIViewController {
     }
     
+    private func updateCreateButtonState() {
+        createButton.isEnabled = isNameEntered && isCategorySelected  && isEmojiSelected && isColorSelected
```
А что если сделать вот такой вариант, запись кажется короче и легче для восприятия, что скажешь ?
```
        let isEnabledButton = isNameEntered && isCategorySelected  && isEmojiSelected && isColorSelected
        createButton.isEnabled = isEnabledButton
        let backgroundColor: UIColor = isEnabledButton ? .ypBlackDay : .ypGray
        createButton.backgroundColor = backgroundColor
```

MiraiChan replied:
Ok!

So1ovyov commented on Tracker/TrackersViewController/TrackersViewController.swift:
```diff
@@ -264,21 +264,28 @@ extension TrackersViewController: UITextViewDelegate {
+            currentCategory.title == category
+        }
+        if foundCategory != nil {
+            self.categories = self.categories.map { currentCategory in
```
Метод кажется нагруженным, давай попробуем декомпозировать его и вынести часть логики в приватный метод

MiraiChan replied:
Done.

So1ovyov commented on Tracker/Onboarding/OnboardingViewController.swift:
```diff
@@ -0,0 +1,162 @@
+        ])
+    }
+    
+    func setupOnboardingPage(imageName: String, labelText: String) -> UIViewController {
```
Неплохое решение, но можно было бы разгрузить этот ViewController, следующим образом:
1. Создать отдельный PageViewController, в который при инициализации ты подставляла нужную картинку/текстовку/ другие настройки
2. Добавить 2 таких VC в данный `OnboardingViewController`

MiraiChan replied:
Done.

So1ovyov commented on Tracker/Category/CategoryViewController.swift:
```diff
@@ -0,0 +1,208 @@
+    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
+        guard let cell = tableView.dequeueReusableCell(withIdentifier: cellReuseIdentifier, for: indexPath) as? CategoryCell else { return UITableViewCell() }
+        
+        if indexPath.row < viewModel.categories.count {
```
Очень много логики во ViewController-е, я рекомендую по максимуму всю возможную бизнес логику вынести во ViewModel

MiraiChan committed:
Changes were made according review:
- aViewController has been created to display a single page;
- a UIViewController has been created, which will be displayed in the UIPageViewController.

MiraiChan committed:
Fixed force-unwrap.

MiraiChan committed:
The overloaded method 'appendTracker' was decomposed and part of the logic was transferred to a private method.
Code was simplified according to reviewer's recommendations.

MiraiChan commented:
Всё, у меня код не падает.

So1ovyov commented on Tracker/Onboarding/PageViewController.swift:
```diff
@@ -0,0 +1,56 @@
+
+import UIKit
+
+class PageViewController: UIViewController {
```
Можно еще добавить модификатор `final`

MiraiChan replied:
ой, забыла

MiraiChan committed:
fatalError replaced with assertionFailure, 'final' added to class PageViewController.
